File: uplc/benches/src/machine/cek.rs
use bumpalo::{collections::Vec as BumpVec, Bump};

use crate::{
    binder::Eval,
    machine::{context::Context, env::Env, state::MachineState},
    term::Term,
};

use super::{
    cost_model::StepKind,
    discharge,
    info::MachineInfo,
    runtime::{BuiltinSemantics, Runtime},
    value::Value,
    CostModel, ExBudget, MachineError,
};

pub struct Machine<'a> {
    pub(super) arena: &'a Bump,
    ex_budget: ExBudget,
    unbudgeted_steps: [u8; 10],
    pub(super) costs: CostModel,
    slippage: u8,
    pub(super) logs: Vec<String>,
    pub(super) semantics: BuiltinSemantics,
}

impl<'a> Machine<'a> {
    pub fn new(
        arena: &'a Bump,
        initial_budget: ExBudget,
        costs: CostModel,
        semantics: BuiltinSemantics,
    ) -> Self {
        Machine {
            arena,
            ex_budget: initial_budget,
            unbudgeted_steps: [0; 10],
            costs,
            slippage: 200,
            logs: Vec::new(),
            semantics,
        }
    }

    pub fn info(self) -> MachineInfo {
        MachineInfo {
            consumed_budget: self.ex_budget,
            logs: self.logs,
        }
    }

    pub fn run<V>(&mut self, term: &'a Term<'a, V>) -> Result<&'a Term<'a, V>, MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        self.spend_budget(ExBudget::start_up())?;

        let initial_context = Context::no_frame(self.arena);

        let mut state =
            MachineState::compute(self.arena, initial_context, Env::new_in(self.arena), term);

        loop {
            let step = match state {
                MachineState::Compute(context, env, term) => self.compute(context, env, term),
                MachineState::Return(context, value) => self.return_compute(context, value),
                MachineState::Done(term) => {
                    return Ok(term);
                }
            };

            state = step?;
        }
    }

    pub fn compute<V>(
        &mut self,
        context: &'a Context<'a, V>,
        env: &'a Env<'a, V>,
        term: &'a Term<'a, V>,
    ) -> Result<&'a mut MachineState<'a, V>, MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        match term {
            Term::Var(name) => {
                self.step_and_maybe_spend(StepKind::Var)?;

                let value = env
                    .lookup(name.index())
                    .ok_or(MachineError::OpenTermEvaluated(term))?;

                let state = MachineState::return_(self.arena, context, value);

                Ok(state)
            }
            Term::Lambda { parameter, body } => {
                self.step_and_maybe_spend(StepKind::Lambda)?;

                let value = Value::lambda(self.arena, *parameter, body, env);

                let state = MachineState::return_(self.arena, context, value);

                Ok(state)
            }
            Term::Apply { function, argument } => {
                self.step_and_maybe_spend(StepKind::Apply)?;

                let frame = Context::frame_await_fun_term(self.arena, env, argument, context);

                let state = MachineState::compute(self.arena, frame, env, function);

                Ok(state)
            }
            Term::Delay(body) => {
                self.step_and_maybe_spend(StepKind::Delay)?;

                let value = Value::delay(self.arena, body, env);

                let state = MachineState::return_(self.arena, context, value);

                Ok(state)
            }
            Term::Force(body) => {
                self.step_and_maybe_spend(StepKind::Force)?;

                let frame = Context::frame_force(self.arena, context);

                let state = MachineState::compute(self.arena, frame, env, body);

                Ok(state)
            }
            Term::Constr { tag, fields } => {
                self.step_and_maybe_spend(StepKind::Constr)?;

                if let Some((first, terms)) = fields.split_first() {
                    let frame = Context::frame_constr_empty(self.arena, env, *tag, terms, context);

                    let state = MachineState::compute(self.arena, frame, env, first);

                    Ok(state)
                } else {
                    let value = Value::constr_empty(self.arena, *tag);

                    let state = MachineState::return_(self.arena, context, value);

                    Ok(state)
                }
            }
            Term::Case { constr, branches } => {
                self.step_and_maybe_spend(StepKind::Case)?;

                let frame = Context::frame_cases(self.arena, env, branches, context);

                let state = MachineState::compute(self.arena, frame, env, constr);

                Ok(state)
            }
            Term::Constant(constant) => {
                self.step_and_maybe_spend(StepKind::Constant)?;

                let value = Value::con(self.arena, constant);

                let state = MachineState::return_(self.arena, context, value);

                Ok(state)
            }
            Term::Builtin(fun) => {
                self.step_and_maybe_spend(StepKind::Builtin)?;

                let runtime = Runtime::new(self.arena, fun);

                let value = Value::builtin(self.arena, runtime);

                let state = MachineState::return_(self.arena, context, value);

                Ok(state)
            }
            Term::Error => Err(MachineError::ExplicitErrorTerm),
        }
    }

    pub fn return_compute<V>(
        &mut self,
        context: &'a Context<'a, V>,
        value: &'a Value<'a, V>,
    ) -> Result<&'a mut MachineState<'a, V>, MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        match context {
            Context::FrameAwaitFunTerm(arg_env, argument, context) => {
                let context = Context::frame_await_arg(self.arena, value, context);

                let state = MachineState::compute(self.arena, context, arg_env, argument);

                Ok(state)
            }
            Context::FrameAwaitArg(function, context) => {
                self.apply_evaluate(context, function, value)
            }
            Context::FrameAwaitFunValue(argument, context) => {
                self.apply_evaluate(context, value, argument)
            }
            Context::FrameForce(context) => self.force_evaluate(context, value),
            Context::FrameConstr(env, tag, terms, values, context) => {
                let mut new_values = BumpVec::with_capacity_in(values.len() + 1, self.arena);

                for value in values.iter() {
                    new_values.push(*value);
                }

                let values = self.arena.alloc(new_values);

                if let Some((first, terms)) = terms.split_first() {
                    let frame =
                        Context::frame_constr(self.arena, env, *tag, terms, values, context);

                    let state = MachineState::compute(self.arena, frame, env, first);

                    Ok(state)
                } else {
                    let value = Value::constr(self.arena, *tag, values);

                    let state = MachineState::return_(self.arena, context, value);

                    Ok(state)
                }
            }
            Context::FrameCases(env, branches, context) => match value {
                Value::Constr(tag, fields) => match branches.get(*tag) {
                    Some(branch) => {
                        let frame = self.transfer_arg_stack(fields, context);

                        let state = MachineState::compute(self.arena, frame, env, branch);

                        Ok(state)
                    }
                    None => Err(MachineError::MissingCaseBranch(branches, value)),
                },
                v => Err(MachineError::NonConstrScrutinized(v)),
            },
            Context::NoFrame => {
                if self.unbudgeted_steps[9] > 0 {
                    self.spend_unbudgeted_steps()?;
                }

                let term = discharge::value_as_term(self.arena, value);

                let state = MachineState::done(self.arena, term);

                Ok(state)
            }
        }
    }

    fn force_evaluate<V>(
        &mut self,
        context: &'a Context<'a, V>,
        value: &'a Value<'a, V>,
    ) -> Result<&'a mut MachineState<'a, V>, MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        match value {
            Value::Delay(term, env) => Ok(MachineState::compute(self.arena, context, env, term)),
            Value::Builtin(runtime) => {
                if runtime.needs_force() {
                    let value = if runtime.is_ready() {
                        self.call(runtime)?
                    } else {
                        Value::builtin(self.arena, runtime.force(self.arena))
                    };

                    let state = MachineState::return_(self.arena, context, value);

                    Ok(state)
                } else {
                    let term = discharge::value_as_term(self.arena, value);

                    Err(MachineError::BuiltinTermArgumentExpected(term))
                }
            }
            rest => Err(MachineError::NonPolymorphicInstantiation(rest)),
        }
    }

    fn apply_evaluate<V>(
        &mut self,
        context: &'a Context<'a, V>,
        function: &'a Value<'a, V>,
        argument: &'a Value<'a, V>,
    ) -> Result<&'a mut MachineState<'a, V>, MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        match function {
            Value::Lambda { body, env, .. } => {
                let new_env = env.push(self.arena, argument);

                let state = MachineState::compute(self.arena, context, new_env, body);

                Ok(state)
            }
            Value::Builtin(runtime) => {
                if !runtime.needs_force() && runtime.is_arrow() {
                    let runtime = runtime.push(self.arena, argument);

                    let value = if runtime.is_ready() {
                        self.eval_builtin_app(runtime)?
                    } else {
                        Value::builtin(self.arena, runtime)
                    };

                    let state = MachineState::return_(self.arena, context, value);

                    Ok(state)
                } else {
                    let term = discharge::value_as_term(self.arena, function);

                    Err(MachineError::UnexpectedBuiltinTermArgument(term))
                }
            }
            rest => Err(MachineError::NonFunctionApplication(argument, rest)),
        }
    }

    fn eval_builtin_app<V>(
        &mut self,
        runtime: &'a Runtime<'a, V>,
    ) -> Result<&'a Value<'a, V>, MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        self.call(runtime)
    }

    fn transfer_arg_stack<V>(
        &mut self,
        fields: &'a [&'a Value<'a, V>],
        context: &'a Context<'a, V>,
    ) -> &'a Context<'a, V>
    where
        V: Eval<'a>,
    {
        if let Some((first, rest)) = fields.split_first() {
            let context = Context::frame_await_fun_value(self.arena, first, context);

            self.transfer_arg_stack(rest, context)
        } else {
            context
        }
    }

    fn step_and_maybe_spend<V>(&mut self, step: StepKind) -> Result<(), MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        let index = step as usize;

        self.unbudgeted_steps[index] += 1;
        self.unbudgeted_steps[9] += 1;

        if self.unbudgeted_steps[9] >= self.slippage {
            self.spend_unbudgeted_steps()?;
        }

        Ok(())
    }

    fn spend_unbudgeted_steps<V>(&mut self) -> Result<(), MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        for step_kind in 0..self.unbudgeted_steps.len() - 1 {
            let mut unspent_step_budget = self.costs.machine_costs.get(step_kind);

            unspent_step_budget.occurrences(self.unbudgeted_steps[step_kind] as i64);

            self.spend_budget(unspent_step_budget)?;

            self.unbudgeted_steps[step_kind] = 0;
        }

        self.unbudgeted_steps[9] = 0;

        Ok(())
    }

    pub(super) fn spend_budget<V>(
        &mut self,
        spend_budget: ExBudget,
    ) -> Result<(), MachineError<'a, V>>
    where
        V: Eval<'a>,
    {
        self.ex_budget.mem -= spend_budget.mem;
        self.ex_budget.cpu -= spend_budget.cpu;

        if self.ex_budget.mem < 0 || self.ex_budget.cpu < 0 {
            Err(MachineError::OutOfExError(self.ex_budget))
        } else {
            Ok(())
        }
    }
}

File: pkg/machine/cek.go
package machine

import (
	"errors"
	"fmt"

	"github.com/blinklabs-io/plutigo/pkg/syn"
)

type Machine struct {
	costs           CostModel
	slippage        uint32
	exBudget        ExBudget
	unbudgetedSteps [10]uint32
	Logs            []string
}

func NewMachine(slippage uint32) Machine {
	return Machine{
		costs:    DefaultCostModel,
		slippage: slippage,
		exBudget: DefaultExBudget,
		Logs:     make([]string, 0),

		unbudgetedSteps: [10]uint32{0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
	}
}

func (m *Machine) Run(term syn.Term[syn.Eval]) (syn.Term[syn.Eval], error) {
	startupBudget := m.costs.machineCosts.startup
	if err := m.spendBudget(startupBudget); err != nil {
		return nil, err
	}

	var state MachineState = Compute{Ctx: NoFrame{}, Env: Env([]Value{}), Term: term}
	var err error

	for {
		switch v := state.(type) {
		case Compute:
			state, err = m.compute(v.Ctx, v.Env, v.Term)
		case Return:
			state, err = m.returnCompute(v.Ctx, v.Value)
		case Done:
			return v.term, nil
		default:
			panic("THIS SHOULD NOT HAPPEN")
		}
		if err != nil {
			return nil, err
		}
	}
}

func (m *Machine) compute(
	context MachineContext,
	env Env,
	term syn.Term[syn.Eval],
) (MachineState, error) {
	var state MachineState

	switch t := term.(type) {
	case *syn.Var[syn.Eval]:
		if err := m.stepAndMaybeSpend(ExVar); err != nil {
			return nil, err
		}

		value, exists := env.lookup(uint(t.Name.LookupIndex()))

		if !exists {
			return nil, errors.New("open term evaluated")
		}

		state = Return{Ctx: context, Value: *value}
	case *syn.Delay[syn.Eval]:
		if err := m.stepAndMaybeSpend(ExDelay); err != nil {
			return nil, err
		}

		value := Delay{
			Body: t.Term,
			Env:  env,
		}

		state = Return{Ctx: context, Value: value}
	case *syn.Lambda[syn.Eval]:
		if err := m.stepAndMaybeSpend(ExLambda); err != nil {
			return nil, err
		}

		value := Lambda{
			ParameterName: t.ParameterName,
			Body:          t.Body,
			Env:           env,
		}

		state = Return{Ctx: context, Value: value}
	case *syn.Apply[syn.Eval]:

		if err := m.stepAndMaybeSpend(ExApply); err != nil {
			return nil, err
		}

		frame := FrameAwaitFunTerm{
			Env:  env,
			Term: t.Argument,
			Ctx:  context,
		}

		state = Compute{
			Ctx:  frame,
			Env:  env,
			Term: t.Function,
		}
	case *syn.Constant:
		if err := m.stepAndMaybeSpend(ExConstant); err != nil {
			return nil, err
		}

		state = Return{
			Ctx: context,
			Value: Constant{
				Constant: t.Con,
			},
		}

	case *syn.Force[syn.Eval]:
		if err := m.stepAndMaybeSpend(ExForce); err != nil {
			return nil, err
		}

		frame := FrameForce{
			Ctx: context,
		}

		state = Compute{
			Ctx:  frame,
			Env:  env,
			Term: t.Term,
		}

	case *syn.Error:
		return nil, errors.New("Eval Failure")

	case *syn.Builtin:
		if err := m.stepAndMaybeSpend(ExBuiltin); err != nil {
			return nil, err
		}

		state = Return{
			Ctx: context,
			Value: Builtin{
				Func:   t.DefaultFunction,
				Args:   []Value{},
				Forces: 0,
			},
		}
	case *syn.Constr[syn.Eval]:
		if err := m.stepAndMaybeSpend(ExConstr); err != nil {
			return nil, err
		}

		fields := *t.Fields

		if len(fields) == 0 {
			state = Return{
				Ctx: context,
				Value: Constr{
					Tag:    t.Tag,
					Fields: []Value{},
				},
			}
		} else {
			first_field := fields[0]

			rest := fields[1:]

			frame := FrameConstr{
				Ctx:            context,
				Tag:            t.Tag,
				Fields:         rest,
				ResolvedFields: []Value{},
				Env:            env,
			}

			state = Compute{
				Ctx:  frame,
				Env:  env,
				Term: first_field,
			}
		}
	case *syn.Case[syn.Eval]:
		if err := m.stepAndMaybeSpend(ExCase); err != nil {
			return nil, err
		}

		frame := FrameCases{
			Env:      env,
			Ctx:      context,
			Branches: *t.Branches,
		}

		state = Compute{
			Ctx:  frame,
			Env:  env,
			Term: t.Constr,
		}
	default:
		fmt.Printf("THING %v\n", state)
		panic("WHAT HAPPENED HERE???")
	}

	return state, nil
}

func (m *Machine) returnCompute(context MachineContext, value Value) (MachineState, error) {
	var state MachineState
	var err error

	switch c := context.(type) {
	case FrameAwaitArg:
		state, err = m.applyEvaluate(c.Ctx, c.Value, value)

		if err != nil {
			return nil, err
		}
	case FrameAwaitFunTerm:
		state = Compute{
			Ctx: FrameAwaitArg{
				Ctx:   c.Ctx,
				Value: value,
			},
			Env:  c.Env,
			Term: c.Term,
		}
	case FrameAwaitFunValue:
		state, err = m.applyEvaluate(c.Ctx, value, c.Value)

		if err != nil {
			return nil, err
		}
	case FrameForce:
		state, err = m.forceEvaluate(c.Ctx, value)

		if err != nil {
			return nil, err
		}
	case FrameConstr:
		resolvedFields := append(c.ResolvedFields, value)

		fields := c.Fields

		if len(fields) == 0 {
			state = Return{
				Ctx: c.Ctx,
				Value: Constr{
					Tag:    c.Tag,
					Fields: resolvedFields,
				},
			}
		} else {
			first_field := fields[0]
			rest := fields[1:]

			frame := FrameConstr{
				Ctx:            context,
				Tag:            c.Tag,
				Fields:         rest,
				ResolvedFields: resolvedFields,
				Env:            c.Env,
			}

			state = Compute{
				Ctx:  frame,
				Env:  c.Env,
				Term: first_field,
			}
		}
	case FrameCases:
		switch v := value.(type) {
		case Constr:
			if indexExists(c.Branches, int(v.Tag)) {
				state = Compute{
					Ctx:  transferArgStack(v.Fields, c.Ctx),
					Env:  c.Env,
					Term: c.Branches[v.Tag],
				}
			} else {
				return nil, errors.New("MissingCaseBranch")
			}
		default:
			return nil, errors.New("NonConstrScrutinized")
		}
	case NoFrame:
		if m.unbudgetedSteps[9] > 0 {
			if err := m.spendUnbudgetedSteps(); err != nil {
				return nil, err
			}
		}

		state = Done{
			term: dischargeValue(value),
		}
	}
	return state, nil
}

func (m *Machine) forceEvaluate(context MachineContext, value Value) (MachineState, error) {

	var state MachineState

	switch v := value.(type) {
	case Delay:
		state = Compute{
			Ctx:  context,
			Env:  v.Env,
			Term: v.Body,
		}
	case Builtin:
		if v.NeedsForce() {
			var resolved Value

			b := v.ConsumeForce()

			if b.IsReady() {
				var err error
				resolved, err = m.evalBuiltinApp(b)
				if err != nil {
					return nil, err
				}
			} else {
				resolved = b
			}

			state = Return{
				Ctx:   context,
				Value: resolved,
			}
		} else {
			return nil, errors.New("BuiltinTermArgumentExpected")
		}
	default:
		return nil, errors.New("NonPolymorphicInstantiation")
	}
	return state, nil
}

func (m *Machine) applyEvaluate(context MachineContext, function Value, arg Value) (MachineState, error) {
	var state MachineState

	switch f := function.(type) {
	case Lambda:
		env := make(Env, len(f.Env))
		copy(env, f.Env)

		env = append(env, arg)

		state = Compute{
			Ctx:  context,
			Env:  env,
			Term: f.Body,
		}
	case Builtin:
		if !f.NeedsForce() && f.IsArrow() {
			var resolved Value

			b := f.ApplyArg(arg)

			if b.IsReady() {
				var err error

				resolved, err = m.evalBuiltinApp(b)
				if err != nil {
					return nil, err
				}
			} else {
				resolved = b
			}

			state = Return{
				Ctx:   context,
				Value: resolved,
			}
		} else {
			return nil, errors.New("UnexpectedBuiltinTermArgument")
		}
	default:
		return nil, errors.New("NonFunctionalApplication")
	}
	return state, nil
}

func transferArgStack(fields []Value, ctx MachineContext) MachineContext {
	c := ctx
	for arg := len(fields) - 1; arg >= 0; arg-- {
		c = FrameAwaitFunValue{
			Ctx:   c,
			Value: fields[arg],
		}
	}
	return c
}

func dischargeValue(value Value) syn.Term[syn.Eval] {
	var dischargedTerm syn.Term[syn.Eval]

	switch v := value.(type) {
	case Constant:
		dischargedTerm = syn.Constant{
			Con: v.Constant,
		}
	case Builtin:
		var forcedTerm syn.Term[syn.Eval]
		forcedTerm = syn.Builtin{
			DefaultFunction: v.Func,
		}

		for i := uint(0); i < v.Forces; i++ {
			forcedTerm = syn.Force[syn.Eval]{
				Term: forcedTerm,
			}
		}

		for _, arg := range v.Args {
			forcedTerm = syn.Apply[syn.Eval]{
				Function: forcedTerm,
				Argument: dischargeValue(arg),
			}
		}

		dischargedTerm = forcedTerm

	case Delay:
		dischargedTerm = syn.Delay[syn.Eval]{
			Term: withEnv(0, v.Env, v.Body),
		}

	case Lambda:
		dischargedTerm = syn.Lambda[syn.Eval]{
			ParameterName: v.ParameterName,
			Body:          withEnv(1, v.Env, v.Body),
		}

	case Constr:
		fields := []syn.Term[syn.Eval]{}

		for _, f := range v.Fields {
			fields = append(fields, dischargeValue(f))
		}

		dischargedTerm = syn.Constr[syn.Eval]{
			Tag:    v.Tag,
			Fields: &fields,
		}
	}

	return dischargedTerm
}

func withEnv(lamCnt uint, env Env, term syn.Term[syn.Eval]) syn.Term[syn.Eval] {
	var dischargedTerm syn.Term[syn.Eval]

	switch t := term.(type) {
	case syn.Var[syn.Eval]:
		if lamCnt >= uint(t.Name.LookupIndex()) {
			dischargedTerm = t
		} else if val, exists := env.lookup(uint(t.Name.LookupIndex()) - lamCnt); exists {
			dischargedTerm = dischargeValue(*val)
		} else {
			dischargedTerm = t
		}

	case syn.Lambda[syn.Eval]:
		dischargedTerm = syn.Lambda[syn.Eval]{
			ParameterName: t.ParameterName,
			Body:          withEnv(lamCnt+1, env, t.Body),
		}

	case syn.Apply[syn.Eval]:
		dischargedTerm = syn.Apply[syn.Eval]{
			Function: withEnv(lamCnt, env, t.Function),
			Argument: withEnv(lamCnt, env, t.Argument),
		}
	case syn.Delay[syn.Eval]:
		dischargedTerm = syn.Delay[syn.Eval]{
			Term: withEnv(lamCnt, env, t.Term),
		}

	case syn.Force[syn.Eval]:
		dischargedTerm = syn.Force[syn.Eval]{
			Term: withEnv(lamCnt, env, t.Term),
		}

	case syn.Constr[syn.Eval]:
		fields := []syn.Term[syn.Eval]{}

		for _, f := range *t.Fields {
			fields = append(fields, withEnv(lamCnt, env, f))
		}

		dischargedTerm = syn.Constr[syn.Eval]{
			Tag:    t.Tag,
			Fields: &fields,
		}
	case syn.Case[syn.Eval]:
		branches := []syn.Term[syn.Eval]{}

		for _, b := range *t.Branches {
			branches = append(branches, withEnv(lamCnt, env, b))
		}

		dischargedTerm = syn.Case[syn.Eval]{
			Constr:   withEnv(lamCnt, env, t.Constr),
			Branches: &branches,
		}
	default:
		dischargedTerm = t

	}

	return dischargedTerm
}

func (m *Machine) stepAndMaybeSpend(step StepKind) error {
	m.unbudgetedSteps[step] += 1
	m.unbudgetedSteps[9] += 1

	if m.unbudgetedSteps[9] >= m.slippage {
		if err := m.spendUnbudgetedSteps(); err != nil {
			return err
		}
	}

	return nil
}

func (m *Machine) spendUnbudgetedSteps() error {
	for i := 0; i < len(m.unbudgetedSteps)-1; i++ {
		unspent_step_budget :=
			m.costs.machineCosts.get(StepKind(i))

		unspent_step_budget.occurrences(m.unbudgetedSteps[i])

		if err := m.spendBudget(unspent_step_budget); err != nil {
			return err
		}

		m.unbudgetedSteps[i] = 0
	}

	m.unbudgetedSteps[9] = 0

	return nil
}

func (m *Machine) spendBudget(exBudget ExBudget) error {
	m.exBudget.mem -= exBudget.mem
	m.exBudget.cpu -= exBudget.cpu

	if m.exBudget.mem < 0 || m.exBudget.cpu < 0 {
		return errors.New("out of budget")
	}

	return nil
}
